### ISSUE TITLE
MinGW-Build broken?

### ISSUE THREAD

**Issue by ArKay74** [nesbox/TIC-80#2163]
I checked out the current git and tried building it with mingw (winlibs 64-bit standalone build with GCC 12.2.0) following the instructions from the README.

The "janet" module doesn't build because it seems to be using a configuration that doesn't work on Windows. First it tries to run "uname" which doesn't exist and then the "cc" compiler (even though GCC was found and configured and the other projects up to that point compiled correctly).

```
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Check for working C compiler: B:/mingw64/bin/gcc.exe
-- Check for working C compiler: B:/mingw64/bin/gcc.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: B:/mingw64/bin/g++.exe
-- Check for working CXX compiler: B:/mingw64/bin/g++.exe - works
```

```
D:\_Development_\C\TIC-80\build>mingw32-make janet
[100%] Generating ../vendor/janet/build/c/janet.c
The system cannot find the path specified.
process_begin: CreateProcess(NULL, uname -s, ...) failed.
makefile:59: pipe: Bad file descriptor
The syntax of the command is incorrect.
process_begin: CreateProcess(NULL, cc -DJANET_BOOTSTRAP -DJANET_BUILD=\"local\" -O0 -g -std=c99 -Wall -Wextra -Isrc/include -Isrc/conf -fvisibility=hidden -fPIC -o build/core/abstract.boot.o -c src/core/abstract.c, ...) failed.
make (e=2): The system cannot find the file specified.
mingw32-make[4]: *** [makefile:164: build/core/abstract.boot.o] Error 2
mingw32-make[3]: *** [CMakeFiles\janet.dir\build.make:81: ../vendor/janet/build/c/janet.c] Error 2
mingw32-make[2]: *** [CMakeFiles\Makefile2:338: CMakeFiles/janet.dir/all] Error 2
mingw32-make[1]: *** [CMakeFiles\Makefile2:346: CMakeFiles/janet.dir/rule] Error 2
mingw32-make: *** [makefile:222: janet] Error 2

```

**Comment by sogaiu** [comment]
May be @NPO-197 or @AlecTroemel might have some insight into this.

There were some chagnes recently, but IIUC, there was a report of success here: https://github.com/nesbox/TIC-80/issues/2154#issuecomment-1496357169

**Comment by borbware** [comment]
i had a similar issue on one machine (winlibs, tried with versions 11, 12 and 13), but on another machine the mingw build just works (winlibs 12.2).

**Comment by nesbox** [comment]
temporary disabled Janet by default here 6b87ce56fe152be1ebcd485300f6563c08adcbaa

**Comment by AlecTroemel** [comment]
> temporary disabled Janet by default here [6b87ce5](https://github.com/nesbox/TIC-80/commit/6b87ce56fe152be1ebcd485300f6563c08adcbaa)

hmm, looks like I forgot to use `if(BUILD_WITH_JANET)` in part of the cmakelist.txt  https://github.com/nesbox/TIC-80/actions/runs/5092851254/jobs/9154743218?pr=2209#step:6:2229

**Comment by AlecTroemel** [comment]
actually its [this reference ](https://github.com/nesbox/TIC-80/blob/main/build/baremetalpi/Makefile#L38) thats killing the pipeline, at least in part.. I honestly have no idea how to use `BUILD_WITH_JANET` in that makefile, or if thats even possible. FWIW it looks like disabling ruby would have this same problem

HOWEVER, this is all unrelated to the broken mingw-build. my impression of the mingw ecosystem is that its pretty fragmented and not the best named. Im fairly confident that at least some form of mingw works with janet. Maybe we just need to update the readme?

**Comment by sogaiu** [comment]
Locally building the janet target is ok for me:
```
$ mingw32-make.exe janet
Generating C:/Users/user/Desktop/TIC-80/vendor/janet/build/c/janet.c
1896699253 2759603 build/c/janet.c
Building C object CMakeFiles/janet.dir/vendor/janet/build/c/janet.c.obj
Linking C static library lib\libjanet.a
Built target janet
```
What's not ok is the default target:
```
$ mingw32-make.exe
[  1%] Building C object CMakeFiles/lua.dir/vendor/lua/lapi.c.obj
[  1%] Building C object CMakeFiles/lua.dir/vendor/lua/lcode.c.obj
[  1%] Building C object CMakeFiles/lua.dir/vendor/lua/lctype.c.obj
[  1%] Building C object CMakeFiles/lua.dir/vendor/lua/ldebug.c.obj
[  1%] Building C object CMakeFiles/lua.dir/vendor/lua/ldo.c.obj
[  1%] Building C object CMakeFiles/lua.dir/vendor/lua/ldump.c.obj
[  2%] Building C object CMakeFiles/lua.dir/vendor/lua/lfunc.c.obj
[  2%] Building C object CMakeFiles/lua.dir/vendor/lua/lgc.c.obj
[  2%] Building C object CMakeFiles/lua.dir/vendor/lua/llex.c.obj
[  2%] Building C object CMakeFiles/lua.dir/vendor/lua/lmem.c.obj
[  2%] Building C object CMakeFiles/lua.dir/vendor/lua/lobject.c.obj
[  2%] Building C object CMakeFiles/lua.dir/vendor/lua/lopcodes.c.obj
[  3%] Building C object CMakeFiles/lua.dir/vendor/lua/lparser.c.obj
[  3%] Building C object CMakeFiles/lua.dir/vendor/lua/lstate.c.obj
[  3%] Building C object CMakeFiles/lua.dir/vendor/lua/lstring.c.obj
[  3%] Building C object CMakeFiles/lua.dir/vendor/lua/ltable.c.obj
[  3%] Building C object CMakeFiles/lua.dir/vendor/lua/ltm.c.obj
[  3%] Building C object CMakeFiles/lua.dir/vendor/lua/lundump.c.obj
[  4%] Building C object CMakeFiles/lua.dir/vendor/lua/lvm.c.obj
[  4%] Building C object CMakeFiles/lua.dir/vendor/lua/lzio.c.obj
[  4%] Building C object CMakeFiles/lua.dir/vendor/lua/lauxlib.c.obj
[  4%] Building C object CMakeFiles/lua.dir/vendor/lua/lbaselib.c.obj
[  4%] Building C object CMakeFiles/lua.dir/vendor/lua/lcorolib.c.obj
[  4%] Building C object CMakeFiles/lua.dir/vendor/lua/ldblib.c.obj
[  5%] Building C object CMakeFiles/lua.dir/vendor/lua/liolib.c.obj
[  5%] Building C object CMakeFiles/lua.dir/vendor/lua/lmathlib.c.obj
[  5%] Building C object CMakeFiles/lua.dir/vendor/lua/loslib.c.obj
[  5%] Building C object CMakeFiles/lua.dir/vendor/lua/lstrlib.c.obj
[  5%] Building C object CMakeFiles/lua.dir/vendor/lua/ltablib.c.obj
[  5%] Building C object CMakeFiles/lua.dir/vendor/lua/lutf8lib.c.obj
[  6%] Building C object CMakeFiles/lua.dir/vendor/lua/loadlib.c.obj
[  6%] Building C object CMakeFiles/lua.dir/vendor/lua/linit.c.obj
[  6%] Linking C static library lib\liblua.a
[  6%] Built target lua
[  7%] Building C object CMakeFiles/lpeg.dir/vendor/lpeg/lpcap.c.obj
[  7%] Building C object CMakeFiles/lpeg.dir/vendor/lpeg/lpcode.c.obj
[  7%] Building C object CMakeFiles/lpeg.dir/vendor/lpeg/lpprint.c.obj
[  7%] Building C object CMakeFiles/lpeg.dir/vendor/lpeg/lptree.c.obj
[  7%] Building C object CMakeFiles/lpeg.dir/vendor/lpeg/lpvm.c.obj
[  7%] Linking C static library lib\liblpeg.a
[  7%] Built target lpeg
[  7%] Building C object CMakeFiles/wren.dir/vendor/wren/src/optional/wren_opt_meta.c.obj
[  7%] Building C object CMakeFiles/wren.dir/vendor/wren/src/optional/wren_opt_random.c.obj
[  7%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_compiler.c.obj
[  8%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_core.c.obj
[  8%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_debug.c.obj
[  8%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_primitive.c.obj
[  8%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_utils.c.obj
[  8%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_value.c.obj
[  8%] Building C object CMakeFiles/wren.dir/vendor/wren/src/vm/wren_vm.c.obj
[  9%] Linking C static library lib\libwren.a
[  9%] Built target wren
[  9%] Creating directories for 'mruby_vendor'
[ 10%] No download step for 'mruby_vendor'
[ 10%] No update step for 'mruby_vendor'
[ 10%] No patch step for 'mruby_vendor'
[ 10%] No configure step for 'mruby_vendor'
[ 10%] Performing build step for 'mruby_vendor'
'C:\msys64\usr\bin\rake' is not recognized as an internal or external command,
operable program or batch file.
mingw32-make[2]: *** [CMakeFiles\mruby_vendor.dir\build.make:85: mruby_vendor-prefix/src/mruby_v
endor-stamp/mruby_vendor-build] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:303: CMakeFiles/mruby_vendor.dir/all] Error 2
mingw32-make: *** [Makefile:135: all] Error 2
```
I don't know why it's looking for `C:\msys64\usr\bin\rake`.

```
$ which rake
/usr/bin/rake

user@DESKTOP MINGW64 /c/Users/user/Desktop/TIC-80/build
$ echo $PATH
/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
```

If I install instead `mingw-w64-x86_64-ruby-rake-compiler` (via `pacman`), I get a `rake` that lives under `/mingw64/bin/rake`, but it doesn't live on my `PATH` and I end up with the same error when trying to build the default target as before.

On a side note, it seems like `WinLibs` is not an "official" MinGW offerring but rather someone else's repackaging -- nothing wrong with that per se, but is it TIC-80's position to try to support that apart from things directly from http://mingw-w64.org/?

**Comment by sogaiu** [comment]
FWIW, disabling MRuby allowed me to successfully build with MinGW (even with Janet) [1].  I did this by:

* `cmake -DBUILD_WITH_MRUBY=FALSE -G "MinGW Makefiles" ..`
* `mingw32-make clean`
* `mingw32-make`

I got a resulting `tic80.exe` file and was able to run both of the Janet demos.

The commit I used for TIC-80 was a82b6cd2d6480ac239e33bc0141adb9e49a27094 - the latest on the main branch at the time of writing this AFAIU.

---

[1] I did see an instance of "compression error" during the generation of `.dat` files, but the build proceeded all the way to the end.

**Comment by sogaiu** [comment]
@NPO-197 When you had success [here](https://github.com/nesbox/TIC-80/issues/2154#issuecomment-1496357169), I'm guessing you didn't have Ruby disabled.  If so, do you recall what sort of setup you used for ruby / rake?

**Comment by nesbox** [comment]
I added some fixes here d256dd3e6c24f377a5b038799e003fda438abf96, but you need **git-bash** installed on your machine.
Maybe later I'll create a more elegant solution :)

**Comment by sogaiu** [comment]
Hurray!  It works here now.

TYVM!
